### PR TITLE
[Portfolio] Directly downloadable persistent certificates in portfolio

### PR DESCRIPTION
**Pull request by legionth**
This PR is the implementation of the requested changes by the JF. Documented in Mantis:
* https://mantis.ilias.de/view.php?id=23761
* https://mantis.ilias.de/view.php?id=24084

Currently a user is only able to add user certificates to its portfolio by creating a verification object in its workspace and add this object to the portfolio. This PR will add a the feature to add and additional option to add a persistent certificate directly into the portfolio, without the creating a verification object.

On selection of an persistent certificate in the GUI form, this feature will create an user certificate (based on the current database enty) in specific folder `<external data directory>/PersistentCertificates`. This behavior is similar to the current process of the verification objects  `<external data directory>/ilVerifification`.
We created the type `crta` to mark this behavior as special from the current handling of the verification objects. Links for the exports HTML and PDF will also be created according to this new feature.

### CODE REVIEW

**Review by mjansenDatabay — APPROVED** [review]

File: Services/COPage/classes/class.ilPCVerificationGUI.php [hunk 1]
```diff
@@ -179,24 +207,47 @@ protected function initForm($a_insert = false)
...
+
+				$userId = $this->user->getId();
+
+				$certificateFileService = new ilPortfolioCertificateFileService();
+				$certificateFileService->createCertificateFile($userId, $objectId);
+
 				$this->content_obj = new ilPCVerification($this->getPage());
 				$this->content_obj->create($this->pg_obj, $this->hier_id, $this->pc_id);
-				$this->content_obj->setData($type, $form->getInput("object"));
+				$this->content_obj->setData('crta', $objectId);
```
Comment (mjansenDatabay) [inline Services/COPage/classes/class.ilPCVerificationGUI.php:265]: @legionth I think we MUST use the 'crta' prefix also for the ID here (to have a globally unique id), but I am not sure. Let's discuss this tomorrow.

Comment (legionth) [reply Services/COPage/classes/class.ilPCVerificationGUI.php:265]: TBH I actually don't know. The code here is very confusing (references in method header, constructors, afterward setters etc.). As far as I can estimate the check for the type before process with the object id, is unique enough.

Comment (legionth) [reply Services/COPage/classes/class.ilPCVerificationGUI.php:265]: So I faked some obj_ids in my table to be the same as other objects in the portfolio (e.g. workspace verification object). It seems setting the type `crta` is unique enough :+1: .

**Comment by mjansenDatabay** [comment]
@alex40724 : Any objections?

**Comment by alex40724** [comment]
@legionth @mjansenDatabay Thank you very much for the PR. The code looks good, but I need to test this before merging. Sorry, but I did not find the time to do this yet. Hopefully until the end of the week.

**Comment by legionth** [comment]
@alex40724 sounds great :+1: I agree, there are lot of changes and it's really hard to review. Take your time & tell us if you have questions :smile:

**Comment by legionth** [comment]
Hey @alex40724, just wanted to check how the review is going. Do you need any additional information from us?

**Review by alex40724 — CHANGES_REQUESTED** [review]
Hi @legionth

Thank you a lot for contributing to ILIAS.

I am sorry that it took some time before I was able to start testing this. As I wrote I wanted not only to have a look at the code but also to test the feature itself a little bit.

Unfortunately I was able to do this very limited, since I am running into errors when trying to configure certificates in this branch, e.g. in courses
"Argument 4 passed to ilCertificateSettingsCourseFormRepository::__construct() must be an instance of ilTemplate"

During testing I experienced at least the following issue:

1.  Create workspace based certificate in portfolio page in current 5.4 branch
2. Moved to patch branch
3. Edit the workspace based certificate in the same Portfolio page

Expected result: The radio option "Workspace Certificates" is selected with the corresponding Certificate being selected in the dropdown.
Result: The radio option "Persistent Certificates" is selected.

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [ ] Please consider to make this PR easier to test by either adressing the bugs in this trunk based branch or moving it to a 5.4.x base.

Please implement the following changes:
- [ ] Please fix the radio options, when editing (issue mentioned above) 
- [ ] After reading the code of `initForm` in `ilPCVerificationGUI` I also expect the `ilSelectInputGUI $persistentObject` won't work correctly when editing, since it's value is never set.

best regards,
Alex

**Comment by legionth** [comment]
Hey @alex40724,

> Unfortunately I was able to do this very limited, since I am running into errors when trying to configure certificates in this branch, e.g. in courses
"Argument 4 passed to ilCertificateSettingsCourseFormRepository::__construct() must be an instance of ilTemplate"

It seems you are on an old trunk to test this feature. This behavior came in through the PR that changed everything to `ilGlobalTemplate` in trunk. These were fixed by: https://github.com/ILIAS-eLearning/ILIAS/pull/1503

To test this properly I recommend the following steps:

* Update the current trunk to the latest version.
* For this step I recommend to add my fork as remote to your test installation e.g. `$ git remote add legionth https://github.com/legionth/ILIAS` (You can of course can use any desired name)
* Checkout the branch `portifolio-certificate`
* You can either test now or use a newer version by rebasing `portifolio-certificate` on `trunk`
* Steps to rebase:
  * `$ git checkout portifolio-certificate`
  * `$ git rebase -i origin/trunk`
* Otherwise you can just merge this branch in to your local trunk

This should be doable without merge conflicts ;)

Beside the general procedure to test this feature, there are currently **bugs in the current `trunk`** because of the change to `ilGlobalTemplate`.

**Comment by legionth** [comment]
> Please implement the following changes:
>     * Please fix the radio options, when editing (issue mentioned above)
>     * After reading the code of `initForm` in `ilPCVerificationGUI` I also expect the `ilSelectInputGUI $persistentObject` won't work correctly when editing, since it's value is never set.

Could you make inline comments which part wouldn't work?

**Comment by legionth** [comment]
>     * Please consider to make this PR easier to test by either addressing the bugs in this trunk based branch or moving it to a 5.4.x base.

I will provide a PR for 5.4 once this PR is accepted for trunk. This PR is based on the current trunk, I rebase continuously to keep this request up to date.

**Comment by alex40724** [comment]
> It seems you are on an old trunk to test this feature.

I added your remote and checkout exactly your branch. If I am not able to test this since it throws errors, I will not accept the PR.

Thanks for rebasing. I will check this version now.

**Comment by alex40724** [comment]
Please fix the editing of certificate components in portfolio pages.

There are several issues:

- Selecting different persistent certificates when editing does not work. ("Edit > Select other Certificate > Save" does not change anything.)
- Editing a certificate will show the wrong certificate (always the first one). See e.g. the screenshot attached (left side the certificate that is being edited / right side the following edit screen with a wrong certificated being selected)
- Workspace certificates are never selected when edited.

bildschirmfoto 2019-02-18 um 17 55 17

**Comment by legionth** [comment]
Thank you for your detailed report :+1: The latest 3 commits will fix these issues.

**Comment by mjansenDatabay** [comment]
@alex40724 Did you already find some time to check the requested bugfixes? We should finalized this before the 5.4.x release.

**Comment by alex40724** [comment]
Sorry to many open ends. I will check this asap.

**Review by alex40724 — APPROVED** [review]
@legionth Ok, thanks again. I tested this successfully. I have got a range-cherry picked version on base of current 5.4 branch locally, which resolved the language conflicts but loses commit history. I could merge this one or you merge this branch yourself, both is ok for me.

**Comment by legionth** [comment]
Resolved the merge conflicts. Ready to merge :shipit:

**Pull request merged**